File: path-c-sqlite/retrieval_check.py
"""Retrieval-only smoke test: no chat model, just the search step.

When a RAG answer is wrong, the first question is always whether the right passage
was retrieved at all. This script answers that question on its own, and it runs in
seconds because no chat model is loaded.

Run:  python retrieval_check.py
"""

from __future__ import annotations

import sys
import time

import models
import rag

# (question, filename fragment that SHOULD be the top hit)
CASES = [
    # Plainly worded - any retriever should get these.
    ("How do I install Foundry Local?", "01-what-is-foundry-local"),
    ("How large should a chunk be?", "05-chunking-strategy"),
    ("How do I store vectors in a database?", "06-sqlite-vector-storage"),
    # Worded differently from the source text - these are the ones that separate
    # semantic search from keyword matching.
    ("The assistant keeps making things up. What do I do?", "07-prompt-engineering"),
    ("Which model is fast enough for students to iterate with?", "08-model-selection"),
    ("Nothing happens when I run the CLI, it just sits there", "12-troubleshooting"),
]

# Deliberately outside the corpus - retrieval will still return its best guess, so
# the scores are what matter here, not the ranking.
OUT_OF_SCOPE = "How many students fit in the lab, and what is the tuition fee?"


def main() -> int:
    conn = rag.connect()
    chunks = rag.load_chunks(conn)
    conn.close()

    if not chunks:
        print("No chunks indexed. Run `python ingest.py` first.", file=sys.stderr)
        return 1

    print(f"Loaded {len(chunks)} chunks. Loading embedding model...\n")
    embedding_model = models.load_embedding_model(on_progress=lambda pct: None)
    client = embedding_model.get_embedding_client()

    doc_embeddings = [c.embedding for c in chunks]
    passed = 0

    for question, expected in CASES:
        started = time.perf_counter()
        query_embedding = client.generate_embedding(question).data[0].embedding
        results = rag.find_relevant(query_embedding, doc_embeddings, top_k=3)
        elapsed = time.perf_counter() - started

        top_source = chunks[results[0][0]].source
        hit_rank = next(
            (r + 1 for r, (i, _) in enumerate(results) if expected in chunks[i].source),
            None,
        )
        ok = hit_rank == 1
        passed += ok

        print(f"{'PASS' if ok else 'MISS'}  {question}")
        print(f"      expected {expected}, top hit {top_source}")
        if hit_rank and not ok:
            print(f"      (expected document was ranked {hit_rank} of 3)")
        for rank, (i, score) in enumerate(results, 1):
            print(f"        {rank}. {score:.3f}  {chunks[i].source}")
        print(f"      retrieved in {elapsed * 1000:.0f} ms\n")

    # Out-of-scope query: report the top score so the grounding threshold is visible.
    query_embedding = client.generate_embedding(OUT_OF_SCOPE).data[0].embedding
    results = rag.find_relevant(query_embedding, doc_embeddings, top_k=3)
    print(f"OUT OF SCOPE  {OUT_OF_SCOPE}")
    print(f"      best score {results[0][1]:.3f} ({chunks[results[0][0]].source})")
    print("      A low score here is what the system prompt has to turn into "
          "an admission of ignorance.\n")

    embedding_model.unload()

    print(f"{passed}/{len(CASES)} questions retrieved the expected document at rank 1.")
    return 0 if passed == len(CASES) else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: path-a-tfidf/retrieval_check.js
/**
 * Retrieval-only smoke test for the TF-IDF path. No chat model is loaded.
 *
 * The question set is identical to path-c-sqlite/retrieval_check.py, so the two
 * outputs can be read side by side. That comparison is the evidence for why the
 * curriculum moves from TF-IDF to embeddings.
 *
 * Usage: node retrieval_check.js
 */
import { config } from "./src/config.js";
import { VectorStore } from "./src/vectorStore.js";

// (question, exact title of the document that SHOULD be the top hit)
const CASES = [
  // Plainly worded - any retriever should get these.
  ["How do I install Foundry Local?", "What Is Foundry Local"],
  ["How large should a chunk be?", "Document Chunking Strategy"],
  ["How do I store vectors in a database?", "Storing Embeddings In SQLite"],
  // Worded differently from the source text - these separate semantic search
  // from keyword matching.
  ["The assistant keeps making things up. What do I do?",
   "Prompt Engineering For Grounded Question Answering"],
  ["Which model is fast enough for students to iterate with?",
   "Choosing A Local Chat Model"],
  ["Nothing happens when I run the CLI, it just sits there",
   "Troubleshooting Guide"],
];

const OUT_OF_SCOPE =
  "How many students fit in the lab, and what is the tuition fee?";

const store = new VectorStore(config.dbPath);
const total = store.count();

if (total === 0) {
  console.error("No chunks indexed. Run `npm run ingest` first.");
  process.exit(1);
}

console.log(`Loaded ${total} chunks (TF-IDF, no embedding model).\n`);

let passed = 0;

for (const [question, expected] of CASES) {
  const started = performance.now();
  const results = store.search(question, 3);
  const elapsed = performance.now() - started;

  const rank = results.findIndex((r) => r.title === expected);
  const ok = rank === 0;
  passed += ok ? 1 : 0;

  console.log(`${ok ? "PASS" : "MISS"}  ${question}`);
  console.log(`      expected "${expected}", top hit "${results[0]?.title}"`);
  if (rank > 0) {
    console.log(`      (expected document was ranked ${rank + 1} of 3)`);
  } else if (rank < 0) {
    console.log(`      (expected document was not in the top 3 at all)`);
  }
  results.forEach((r, i) => {
    console.log(`        ${i + 1}. ${r.score.toFixed(3)}  ${r.title}`);
  });
  console.log(`      retrieved in ${elapsed.toFixed(0)} ms\n`);
}

const oos = store.search(OUT_OF_SCOPE, 3);
console.log(`OUT OF SCOPE  ${OUT_OF_SCOPE}`);
console.log(
  `      best score ${oos[0] ? oos[0].score.toFixed(3) : "n/a"} ` +
    `(${oos[0] ? oos[0].title : "no hit"})\n`
);

console.log(
  `${passed}/${CASES.length} questions retrieved the expected document at rank 1.`
);
store.close();
